_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,883 @@
+              .withPasswordLength(6)
+              .withAlgorithm(HMACAlgorithm.SHA256)
+              .build()
+```
+
+Here we instantiate a new `HOTPGenerator` class, pass it the `secret`, password length as 6, and algorithm type as `SHA256`. We can now use `hotp` to generate the code.
+
+```scala
+val counter = 5
+val code = hotp.generate(counter)
+```
+
+In the code above, we initialized the counter to 5 and used that to generate the `hotp` code.
+
+When the server receives this code, it can verify it in the following way.
+
+```scala
+val isValid = hotp.verify(code, counter)
+```
+
+Remember the client and server must keep their counters in sync for the verification phase to work otherwise resynchronization is needed.
+
+### 2.2 Time-based One Time Password (TOTP)
+
+The TOTP token is generated similarly to HOTP with the main difference being the moving factor. Here the moving factor is based on a time counter. The time counter is calculated by dividing the current Unix time by a timestep value which is the life span of a TOTP usually 30 seconds.
+
+**TOTP scala implementation**
+
+Otp-java also provides an implementation for TOTP token generation.
+
+```scala
+import java.time.Duration
+
+val secret = SecretGenerator.generate()
+
+val totp = new TOTPGenerator.Builder(secret)
+    .withHOTPGenerator(builder => {
+            builder.withPasswordLength(6)
+            builder.withAlgorithm(HMACAlgorithm.SHA256)
+    })
+    .withPeriod(Duration.ofSeconds(30))
+    .build()
+```
+
+Here the `TOTPGenerator` class is instantiated with a `secret`, the `withHOTPGenerator()` method gives access to a `builder` that we use to set the password length and algorithm type. The last method is `withPeriod()` where we set the duration to 30 seconds, which is the lifespan for each `TOTP`.
+
+```scala
+val codeValue = totp.now()
+```
+
+The `now()` method on the `TOTPGenerator` class is used to generate the token.
+In order to verify totp tokens, we also use the `verify()` method.
+
+```scala
+totp.verify(code)
+```
+
+## 3. Two Factor Authentication (2FA)
+
+Two Factor Authentication is an authentication method that requires a user to provide two distinct forms of identification to access a website or resource. Here's how it works.
+
+1. When a user tries to perform a transaction or action on a system, he or she will present some credentials like an email or a phone number.
+1. The system will send a temporary secure PIN-code or token to the user by email or phone number valid for only that session.
+1. When the PIN-code is presented to the system and verified, access will be granted to the user otherwise it will be denied.
+
+The code or token is the One Time Password that we generated in the previous sections
+
+### 3.1 2FA implementation in Scala
+
+In this section, we will create a small application to showcase 2FA using java-otp, Http4s, SendGrid, Google Zxing, and Google Authenticator.
+
+**Creating the User case class.**
```

Comment:
Can remove

Reply:
I removed the section in bold text, not sure if you wanted the text before that removed as well